src/Common/OptimizedRegularExpression.cpp

```diff
@@ -342,6 +342,23 @@ OptimizedRegularExpressionImpl<thread_safe>::OptimizedRegularExpressionImpl(cons
     }
 }
 
+template <bool thread_safe>
+OptimizedRegularExpressionImpl<thread_safe>::OptimizedRegularExpressionImpl(OptimizedRegularExpressionImpl && rhs)
+    : is_trivial(rhs.is_trivial)
+    , required_substring_is_prefix(rhs.required_substring_is_prefix)
+    , is_case_insensitive(rhs.is_case_insensitive)
+    , required_substring(std::move(rhs.required_substring))
+    , re2(std::move(rhs.re2))
+    , number_of_subpatterns(rhs.number_of_subpatterns)
+{
+    if (!required_substring.empty())
+    {
```

Comment:
>What is happening in the latter case is move assignment from a temporary r-value to an l-value. The default move assignment operator of OptimizedRegularExpressionImpl moves all its members. In particular, we move OptimizedRegularExpressionImpl::required_substring and the StringSearcher members of OptimizedRegularExpressionImpl. But my understanding is that after this operation, the pointers inside the StringSearchers still point to the correct required_substring. The situation would be different for a copy operation where the original (copied-from) object goes out of scope after copy - in that case, the pointers would point into nirvana. But again, we are doing a move here which should in theory be safe, right?

If std::string does not have optimization that allows to store first X (32 AFAIR) bytes in the object itself then you were right, however:
- it has such optimization and so this string was stored on stack and later goes out of scope.
- it is not great to rely on this anyway, move can copy

>Implementing respectively disallowing copy or move constructors should always be done in pair with the corresponding copy or move assignment operation. E.g. this here is about assignment but only constructors were fixed.

Compiler should not add move assignment operator if you have non default move constructor

>The situation is not immediately clear to the reader, a comment next/above the StringSearcher members would be nice, e.g. /// StringSearchers store pointers to required_substring, needs care when copying OptimizedRegularExpressionImpl

Actually I want to keep this patch small since this is a bug fix and it should be backported, however I'm fine to slightly depart from this rule.

Reply:
Thanks, that was quite illuminating.

Just as a final question: Conceptually, `const Regexps::Regexp regexp = Regexps::createRegexp<false, false, false>(pattern);` is still a move assignment (even if it by default copies under the hood due to SSO). In this PR, you provide a move constructor. According to your explanation and the C++ standard, that prevents the compiler from generating an implicit move assignment op (--> move assignment is implicitly deleted). In the absence of a move assignment op, the compiler falls back to the copy assignment op but according to [here](https://en.cppreference.com/w/cpp/language/copy_assignment)

>  An implicitly-declared copy assignment operator for class T is defined as deleted if any of the following is true:
>  - T has a user-declared move constructor; 
>  - T has a user-declared move assignment operator. 

that one will also be (implicitly) deleted ^^. So I wonder to which operation `const Regexps::Regexp regexp = Regexps::createRegexp<false, false, false>(pattern);` eventually maps to now?